feat: add support for rendering text with props

brandon1024 opened the pull request:
Introduce the ability to render text with attached text properties. Renderers may now return list of dictionaries where each line of text can include text properties for additional text styling capabilities. This change is backwards compatible with existing renderers; they will continue to work as before.

The semantics of this feature was inspired by popup_create(), which can accept a list of dictionaries with a 'text' and 'props' key.

This feature enhances the capabilities of custom renderers. In particular, it's possible to implement renderers that render content with dynamically-defined style. The screenshot below shows a simple custom renderer that renders vim files in green, config files in grey and others in orange.

![image](https://user-images.githubusercontent.com/22732449/219378808-bbffaf55-0425-4c53-940e-50dcd1f14ac4.png)

Here's a renderer I built using these features: [brandon1024/fern-renderer-nf.vim](https://github.com/brandon1024/fern-renderer-nf.vim/blob/main/autoload/fern/renderer/nf.vim)

A very dumb renderer example:

```vim
" Build and return the renderer.
function! fern#renderer#nf#new() abort
        return {
		\ 'render': funcref('s:render'),
		\ 'index': { lnum -> lnum - 1 },
		\ 'lnum': { index -> index + 1 },
		\ 'syntax': { -> v:null },
		\ 'highlight': { -> v:null }
	\ }
endfunction

" Render given nodes and return a list of lines with text properties.
function! s:render(nodes) abort
	let l:base = len(a:nodes[0].__key)

	if empty(prop_type_get('test'))
		call prop_type_add('test', { 'highlight': 'Normal' })
	endif

	return map(copy(a:nodes), { i, node ->
		\ {
			\ 'text': '>' . repeat(' ', len(node.__key) - l:base) . node.label,
			\ 'type': [{ 'col': 1, 'length': 1, 'type': 'test' }]
		\ }})
endfunction
```

lambdalisue reviewed (commented):
Not sure but does this work on Neovim as well?

brandon1024 commented:
@lambdalisue I don't believe so. From what I've read, [extmarks](https://neovim.io/doc/user/api.html#extmarks) is the neovim-equivalent to textprops but they are not compatible. [This issue](https://github.com/neovim/neovim/issues/13661) describes why textprops weren't introduced in neovim.

I guess the next question is where we go from here :sweat_smile: We might be able to build a shim that uses textprops for vim and extmarks for neovim, so that these features can be used for both editors. Alternatively we could document that textprops are only supported in vim. What are your thoughts? I'd love to support both but I have literally no experience with Neovim, so I'd need some guidance. Documenting that textprops are only supported for Vim seems the easiest to me.

lambdalisue commented:
> Alternatively we could document that textprops are only supported in vim

It sounds reasonable enough. It's OK if this change does not break current features in Neovim (I guess the current code breaks features on Neovim, right?)

brandon1024 commented:
> It sounds reasonable enough.

Cool, this makes my life easier. I'm sure someone will be interested in implementing this for extmarks at some point too.

> It's OK if this change does not break current features in Neovim (I guess the current code breaks features on Neovim, right?)

Existing renderers should continue to work as before, for both vim and neovim :smiley:  The [changes I've introduced](https://github.com/lambdalisue/fern.vim/pull/465/files#diff-50c56ae0ae12ca3441013c63534702db02e4a3dcf02d888f3257ff131aecdc1bR18) check to see if it was passed a list of strings or a list of dictionaries. If passed dictionaries, it would fail, but that would be user error. I can also add a check to see if we are in vim or neovim and throw a useful error message too, or log a message. I might do that actually, not a bad idea.

brandon1024 commented:
I've adjusted the docs, indicating that text properties are only supported in vim. I also added a check and a useful error message when trying to render text properties in neovim.

I guess the next question for you @lambdalisue , would you prefer that we fail abruptly if trying to use text properties in neovim, or just skip the text properties and render a plain-text tree? In code:

```vim
    if !empty(props) && !exists('*prop_add')
      throw 'text properties are not supported, prop_add() does not exist.'
    endif

    for prop in props
      let prop.bufnr = a:bufnr
      call prop_add(lnum + 1, prop.col, prop)
    endfor
```

or

```vim
    if exists('*prop_add')
      for prop in props
        let prop.bufnr = a:bufnr
        call prop_add(lnum + 1, prop.col, prop)
      endfor
    endif
```

I've thought about it a bit, I think the second one makes more sense, since we'd still render a file tree but it just wouldn't be colorized.

lambdalisue commented:
> or just skip the text properties and render a plain-text tree

Seems nice. 

Additionally, I just moved so I need to find time to check PRs so please be patient. I'm sorry for that.

brandon1024 commented:
No worries :-)

brandon1024 commented:
@lambdalisue I think this is ready to go now, whenever you get a second :-)

lambdalisue commented on autoload/fern/internal/buffer.vim:
```diff
@@ -5,14 +5,33 @@ function! fern#internal#buffer#replace(bufnr, content) abort
 endfunction
 
+function! s:replace_buffer_content(bufnr, content) abort
+  for lnum in range(len(a:content))
+    let line = a:content[lnum]
+    let [text, props] = type(line) is# v:t_dict
+      \ ? [line.text, get(line, 'props', [])]
+      \ : [line, []]
+
+    call setbufline(a:bufnr, lnum + 1, text)
+
+    if exists('*prop_add')
+      for prop in props
+        let prop.bufnr = a:bufnr
+        call prop_add(lnum + 1, prop.col, prop)
+      endfor
+    endif
+  endfor
```
The use of for loops in Vim scripts has a significant impact on performance, so this part of the script should be optionally controllable.

brandon1024 commented on autoload/fern/internal/drawer/hover_popup.vim:
```diff
@@ -74,7 +74,7 @@ function! s:show() abort
     call setbufline('%', 1, line)
     call helper.fern.renderer.syntax()
     call helper.fern.renderer.highlight()
-    syntax clear FernRoot
```
I'm pretty sure this is a bug. It was giving me some trouble so I fixed it here :-)

lambdalisue reviewed (approved):
LGTM

lambdalisue commented:
Thanks for your contribution 👍

brandon1024 commented:
Thanks @lambdalisue :-)
